### ISSUE TITLE
Move some module functions to be methods of structs

### ISSUE THREAD

**Issue by theamazingwaffle** [ggez/ggez#748]
Sorry if i'm wrong. I never did serious programming and i'm afraid of it like i'm afraid of going into a dark forest at night.
Also sorry if i'm being unprofessional.

-----

I'm currently (kind of) learning the principles of clean code.
While trying to do some stuff with ggez i thought that passing context is not that intuitive - you often forget that you need to pass it and when you do, you always open the documentation just to make sure the order is right. Ideally all of the functions that need the context would be its methods.
That's because less arguments is better and makes your function easier to understand.

-----
Examples:

Its working like this:
```
fn draw(&mut self, ctx: &mut ggez::Context) -> ggez::GameResult {
  // --snip--
  graphics::draw(ctx, &circle, (na::Point2::new(0.0, 0.0),))?;
  graphics::present(ctx)?;
}
```
But wouldn't it be more intuitive for it to work like this:
```
fn draw(&mut self, ctx: &mut ggez::Context) -> ggez::GameResult {
  // --snip--
  context.draw( &circle, (na::Point2::new(0.0, 0.0),) )?;
  context.present()?;
}
```

Notice how the `fn draw(&mut self, ctx: &mut Context) -> GameResult` still requires the context to be passed to it, but it only happens once. (although i guess uncle bob would still say, with a revolting voice, "Its not clean enough" ¯\_(ツ)_/¯)

Also from your example:
Why use `event::run(...)` and passing the event loop to it, when we can make the run function a part of the event loop?
```
use ggez::event;
// importing all the other libs

fn main() -> GameResult {
  // --snip--
  event::run(&mut context, &mut events_loop, &mut state)?;
}
```
In the example above we needed to import ggez::event (i hate it when there are 'ggez::' all over the place). Also we had to pass the event and context, we have no way of knowing the order of those two. Now lets compare it to this:
```
// not importing ggez::event
fn main() -> GameResult {
  // --snip--
  events_loop.run(&mut context, &mut state)?;
  // way easier to read and way easier to understand what its doing
}
```

(i'm bad with names from the field of solution but bare with me)
Why not take it a step further and make ContextBuilder return some kind of GameObject:
```
  struct GameObject {
    context: Context,
    events_loop: EventsLoop,
  }
```
Then when we need to run it we do:
```
game_object.run(&mut state)
```

Thats it. We just build our game_object with the new `GameObjectBuilder` and we've simplified the code a lot.

i'm so sorry if i said it all wrong, i don't know what i'm talking about i never even created a game (outside of a tiny project in love2d that i'd prefer not to talk about)

So yeah that's my speech, it turned out a lot longer than i expected but i'm proud of it so you better crush my confidence as soon as you can with something like "rust doesn't allow that".

But if this idea was actually great than you can go wild and apply this thing to everything (no, really)
I haven't finished the "clean code" book. If you haven't read it (which i doubt soooo much) i would maybe possibly recommend you to. (?) idk.

**Comment by icefoxen** [comment]
That's been brought up a few times in various places, it's a reasonable question.  There's a few different reasons why things are the way they are.  I thought there had been an issue discussing this before, but I can't find it (though there's a [related issue](https://github.com/ggez/ggez/issues/570) or two) so we might as well do it now.

Why everything is functions instead of methods on `Context`:

 * It's a closer approximation of the LÖVE API -- this is the big one, when I started this I was a total Rust newbie and didn't know a huge amount about game engines either, so I wanted to just stick with an existing pattern.
 * It makes it easier for me as maintainer to know what is part of the stable API and what isn't.  Everything in `ggez` that a user can touch is (or should be) either a data object like an `Image`, a method on one of those objects, or a standalone function.  `Context` and the various sub-pieces of it are implementation details that may be public, but are subject to change.

In general, ggez currently uses a functional-programming-ish structure: everything is functions in modules, all state gets passed into those functions.  The exceptions are things like `Image` or `Font` where their methods mostly affect *only* those objects.  The state is one big honkin' global-ish object, which *could* indeed be broken down into multiple smaller objects, but that would make the API more complicated -- for example `Image::new()` loads things from the filesystem, so you'd have to pass a `FilesystemContext` into it along with the `GraphicsContext`.  That just means more state for the user to juggle, and they would most likely end up just making a `GgezContext` in their actual game and stuffing everything into that anyway.

Having functions in modules also makes it easier to sort things nicely, the "everything is a method on an object" approach is a false dichotomy, and one of the biggest fuckups of the object-oriented programming idea.  If everything were a method, we'd end up with a `Context` that has a bajillion methods on it.  So to understand it we'd have to break things down into sub-contexts, as ggez already does internally, but that brings its own problems.  Would the current `Image::new()` become a method on `Image` as it is now, or a method on `GraphicsContext`, or a method on `FilesystemContext`?  No matter which you choose, someone somewhere would expect it to be somewhere else, and would *ask questions or complain when they couldn't find it in the docs where they expected*.  As it is, all a new and unfamiliar user has to say is "well it's probably somewhere in `ggez::graphics`" and that's where they find it.  You would also probably still end up with lone functions here and there that *aren't* attached to a `Context`, or don't actually use the `Context`, and that gets weird too.

This is all, in the end, a matter of taste.  I've made it the way it is because it makes sense to me.  These *are* indeed decisions involving tradeoffs that would probably start tilting the other way if ggez became larger and more complicated, but *I don't intend to do that*.  ggez's goal is to stay small and simple, so that you hold basically the entire API in your head at once, and the current design reflects that.  Modularity is our best tool to divide large complicated things into small simpler things, but it also has a cost.  Pursuing it for its own sake just gets you [ravioli code](https://en.wikipedia.org/wiki/Spaghetti_code#Ravioli_code).

So, this sort of structure does not preclude good design.  If a function or method takes a `Context`, its always the first argument.  Modules are organized so that you can just do `use ggez::*;` in every file and you will get access to all submodules but won't get name clashes; the only types that imports are the ones you use everywhere, `Context`, `GameError` and `GameResult` (and `ContextBuilder` I guess which I don't have a better place to put).  There's various other small things that try to make ggez easier to use: `DrawParam` is always the last argument in any function that takes it, if I recall correctly.  Types exposed from other crates are always re-exported.  Names for things and setters/getters are consistent.  Stuff like that.

TLDR: I've made it the way it is on purpose, and "methods on objects" and "organized well" are separate things.

**Comment by 17cupsofcoffee** [comment]
>  That said, there's other ways of doing it that are perfectly valid, so there's other game engines with broadly similar goals to ggez

FWIW, pre 0.1, my engine originally had a lot more stuff as methods on `Context`/multiple context objects, and I ended up moving to free functions for very similar reasons as you.

I still think there's probably some hypothetical object-y API that'd work (`coffee`'s rendering API is an interesting approach), but I haven't been able to come up with one that makes me happy yet.

**Comment by theamazingwaffle** [comment]
I should probably already shut up, but  
>  ggez's goal is to stay small and simple, so that you hold basically the entire API in your head at once

Thats the problem. The bigger the library, the more you have to memorize, to keep in your head.  
As uncle bob said:  
> A smart programmer is creating complicated code, because he is smart, but a professional programmer would create simple code that does what you expected.

I imagine a context being like the window, so i guess i'll draw into the context. What would i expect from a window? I wouldn't pass it, i'd draw into it like `context.draw()`. Its just an abstraction, a simplification, i don't say "AAARGH CODE SHOULD BE MODULAR BLAARGH", no i mean that'd make it more intuitive.
